packages/react-native-codegen/src/parsers/typescript/parseTopLevelType.js

```diff
@@ -0,0 +1,177 @@
+      } else {
+        return value;
+      }
+    case 'TSTypeAliasDeclaration':
+      return getValueFromTypes(value.typeAnnotation, types);
+    default:
+      return value;
+  }
+}
+
+function isNull(t: $FlowFixMe) {
+  return (
+    t.type === 'TSNullKeyword' ||
+    t.type === 'TSUndefinedKeyword' ||
+    t.type === 'TSVoidKeyword'
+  );
+}
+
+function handleUnionAndParen(
+  type: $FlowFixMe,
+  result: TopLevelTypeInternal,
+  knownTypes?: TypeDeclarationMap,
+): void {
+  switch (type.type) {
+    case 'TSParenthesizedType': {
+      handleUnionAndParen(type.typeAnnotation, result, knownTypes);
+      break;
+    }
+    case 'TSUnionType': {
+      // the order is important
+      // result.optional must be set first
+      for (const t of type.types) {
+        if (isNull(t)) {
+          result.optional = true;
+        }
+      }
+      for (const t of type.types) {
+        if (!isNull(t)) {
+          handleUnionAndParen(t, result, knownTypes);
+        }
+      }
+      break;
+    }
+    case 'TSTypeReference':
+      if (type.typeName.name === 'Readonly') {
+        handleUnionAndParen(type.typeParameters.params[0], result, knownTypes);
+      } else if (type.typeName.name === 'WithDefault') {
+        if (result.optional) {
+          throw new Error(
+            'WithDefault<> is optional and does not need to be marked as optional. Please remove the union of undefined and/or null',
+          );
+        }
+        if (type.typeParameters.params.length !== 2) {
+          throw new Error(
+            'WithDefault requires two parameters: type and default value.',
+          );
+        }
+        if (result.defaultValue !== undefined) {
+          throw new Error(
+            'Multiple WithDefault is not allowed in a union type.',
+          );
+        }
+        result.optional = true;
+        handleUnionAndParen(type.typeParameters.params[0], result, knownTypes);
+
+        const valueType = type.typeParameters.params[1].type;
+        if (valueType === 'TSLiteralType') {
+          const literal = type.typeParameters.params[1].literal;
+          if (
+            literal.type === 'Literal' ||
+            literal.type === 'StringLiteral' ||
+            literal.type === 'NumericLiteral' ||
+            literal.type === 'BooleanLiteral'
+          ) {
+            if (
+              typeof literal.value === 'string' ||
+              typeof literal.value === 'number' ||
+              typeof literal.value === 'boolean'
+            ) {
+              result.defaultValue = literal.value;
+            }
+          } else if (
+            literal.type === 'UnaryExpression' &&
+            literal.argument.type === 'Literal' &&
+            typeof literal.argument.literal === 'number'
+          ) {
+            result.defaultValue = -literal.argument.value;
+          }
+        } else if (
+          valueType === 'TSNullKeyword' ||
+          valueType === 'TSUndefinedKeyword'
```

Comment:
can't we use `'isNull` here?

Reply:
`isNull` also checked `void` but I don't think `void` would be a default value in `WithDefault`?

Reply:
I finally realized why the original implementation tested `null` and `void`, because Flow's `void` is TypeScript's `undefined` and `void` combined together.